<a href="https://colab.research.google.com/github/AM0ya/Clase_Programacion/blob/main/Actividad_Calificada_07_Datos_Duplicados_269712.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Alfredo Moya Orozco |
| **Matrícula** |269712 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [2]:
# Escribe tu código aquí

df_duplicados_first = hoteles.duplicated(keep='first')
print('Duplicados keep first:Representa la duplicidad de datos marcando primera coincidencia como False.(podria entenderse como un escaneo de arriba hacia abajo)\n',df_duplicados_first.sum(),' Filas duplicadas\n')

df_duplicados_last = hoteles.duplicated(keep='last')
print('Duplicados keep last:Represena la duplicidad de datos, marcando la primera coincidencia como True. (podria entenderse como un escaneo de abajo hacia arriba)\n',df_duplicados_last.sum(),' Filas duplicadas\n')

df_duplicados_false = hoteles.duplicated(keep=False)
print('Duplicados keep False:Marca todos los datos duplicados sin excepcion como True, y deja en False solo los que no estan duplicados.\n',df_duplicados_false.sum(),' Filas duplicadas\n')


Duplicados keep first:Representa la duplicidad de datos marcando primera coincidencia como False.(podria entenderse como un escaneo de arriba hacia abajo)
 31994  Filas duplicadas

Duplicados keep last:Represena la duplicidad de datos, marcando la primera coincidencia como True. (podria entenderse como un escaneo de abajo hacia arriba)
 31994  Filas duplicadas

Duplicados keep False:Marca todos los datos duplicados sin excepcion como True, y deja en False solo los que no estan duplicados.
 40165  Filas duplicadas



### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [3]:
# Escribe tu código aquí


# 1. Cálculo del porcentaje de filas sobrantes respecto al total
total_filas = len(hoteles)
filas_sobrantes = df_duplicados_first.sum()
porcentaje_sobrantes = (filas_sobrantes / total_filas) * 100

# 2. Cálculo del número de grupos repetidos (combinaciones únicas que aparecen más de una vez)
# Usamos value_counts() para contar las ocurrencias de cada combinación idéntica de filas
conteo_combinaciones = hoteles.value_counts(dropna=False)
numero_grupos_repetidos = (conteo_combinaciones > 1).sum()

# Impresión de resultados con formato claro
print(f"Porcentaje de filas sobrantes: {porcentaje_sobrantes:.2f}%")
print(f"Número de grupos repetidos   : {numero_grupos_repetidos}")


Porcentaje de filas sobrantes: 26.80%
Número de grupos repetidos   : 8171


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [4]:
# Escribe tu código aquí

# Creamos una copia del DataFrame y agregamos una columna booleana de duplicados (keep='first')
hoteles_dup = hoteles.copy()
hoteles_dup['es_duplicado'] = hoteles_dup.duplicated(keep='first')

# Agrupamos por hotel para contar reservas totales y filas sobrantes
resumen_hotel = hoteles_dup.groupby('hotel').agg(
    total_reservas=('is_canceled', 'size'),
    filas_sobrantes=('es_duplicado', 'sum')
)

# Calculamos el porcentaje de duplicados respecto al total de cada hotel
resumen_hotel['porcentaje'] = (resumen_hotel['filas_sobrantes'] / resumen_hotel['total_reservas']) * 100

print(resumen_hotel)

              total_reservas  filas_sobrantes  porcentaje
hotel                                                    
City Hotel             79330            25902   32.650952
Resort Hotel           40060             6092   15.207189


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.



**Respuesta 1.4:**

Respuesta: El problema de duplicados en este dataset es considerablemente grave, representando el 26.80% del total de registros (31,994 filas sobrantes de 119,390). Sin embargo, no afecta por igual a ambos hoteles. El City Hotel concentra la gran mayoría del problema con aproximadamente un 34.6% de duplicados (unas 26,000 filas sobrantes), mientras que el Resort Hotel presenta una proporción mucho menor, rondando el 15.0% (unas 6,000 filas sobrantes).

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [5]:
# Escribe tu código aquí

# Filtramos las filas involucradas en duplicados exactos (keep=False)
duplicados_todos = hoteles[hoteles.duplicated(keep=False)]

# Columnas solicitadas en la instruccion
cols_21 = [
    'hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
    'arrival_date_day_of_month', 'adults', 'country', 'market_segment',
    'deposit_type', 'adr', 'reservation_status'
]

# Ordenamos por hotel, tipo de antelacion y fecha de llegada.
# Explicacion: Este ordenamiento agrupa de manera logica las reservas identicas
# para verificar visualmente que las copias coinciden exactamente en sus atributos clave.
duplicados_ordenados = duplicados_todos.sort_values(
    by=['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month']
)

print(duplicados_ordenados[cols_21].head(12))

            hotel  lead_time  arrival_date_year arrival_date_month  \
40643  City Hotel          0               2015             August   
40644  City Hotel          0               2015             August   
40772  City Hotel          0               2015             August   
40802  City Hotel          0               2015             August   
40821  City Hotel          0               2015             August   
40838  City Hotel          0               2015             August   
76792  City Hotel          0               2015             August   
76793  City Hotel          0               2015             August   
76794  City Hotel          0               2015             August   
41026  City Hotel          0               2015             August   
41067  City Hotel          0               2015             August   
41071  City Hotel          0               2015             August   

       arrival_date_day_of_month  adults country market_segment deposit_type  \
40643    

### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [6]:
# Escribe tu código aquí

# Conteo de frecuencia de cada combinacion de filas completas, considerando NaN
conteo_filas = hoteles.value_counts(dropna=False)

# 1. Contar cuantos grupos hay de cada tamaño (mostrando los 10 más frecuentes)
tamanos_grupos = conteo_filas.value_counts().sort_index(ascending=False)
print("10 tamaños de grupos más frecuentes (número de copias y cuántos grupos existen):")
print(tamanos_grupos.head(10))

# 2. Tamaño del grupo mas grande y su combinacion de valores
tam_max = conteo_filas.max()
print(f"\nTamaño del grupo más grande: {tam_max} copias idénticas.")

comb_mas_grande = conteo_filas[conteo_filas == tam_max].index[0]
print("\nCombinación de valores del grupo más grande:")
print(comb_mas_grande)

10 tamaños de grupos más frecuentes (número de copias y cuántos grupos existen):
count
180    1
150    1
109    1
101    1
100    1
99     3
89     1
85     3
84     2
79     3
Name: count, dtype: int64

Tamaño del grupo más grande: 180 copias idénticas.

Combinación de valores del grupo más grande:
('City Hotel', np.int64(1), np.int64(277), np.int64(2016), 'November', np.int64(46), np.int64(7), np.int64(1), np.int64(2), np.int64(2), np.float64(0.0), np.int64(0), 'BB', 'PRT', 'Groups', 'TA/TO', np.int64(0), np.int64(0), np.int64(0), 'A', 'A', np.int64(0), 'Non Refund', np.float64(nan), np.float64(nan), np.int64(0), 'Transient', np.float64(100.0), np.int64(0), np.int64(0), 'Canceled', '2016-04-04')


### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [7]:
# Escribe tu código aquí

# Comparamos el dataset completo frente a las filas involucradas en duplicados (keep=False)
duplicados_todos = hoteles[hoteles.duplicated(keep=False)]

print("--- Distribución de market_segment ---")
print("Dataset completo:")
print(hoteles['market_segment'].value_counts(normalize=True))
print("\nDuplicados (keep=False):")
print(duplicados_todos['market_segment'].value_counts(normalize=True))

print("\n--- Distribución de deposit_type ---")
print("Dataset completo:")
print(hoteles['deposit_type'].value_counts(normalize=True))
print("\nDuplicados (keep=False):")
print(duplicados_todos['deposit_type'].value_counts(normalize=True))

print("\n--- Tasa de cancelación (is_canceled) ---")
print(f"Dataset completo : {hoteles['is_canceled'].mean():.4f}")
print(f"Duplicados (keep=False): {duplicados_todos['is_canceled'].mean():.4f}")

--- Distribución de market_segment ---
Dataset completo:
market_segment
Online TA        0.473046
Offline TA/TO    0.202856
Groups           0.165935
Direct           0.105587
Corporate        0.044350
Complementary    0.006223
Aviation         0.001985
Undefined        0.000017
Name: proportion, dtype: float64

Duplicados (keep=False):
market_segment
Groups           0.416109
Offline TA/TO    0.303996
Online TA        0.205876
Corporate        0.036699
Direct           0.034981
Complementary    0.001867
Aviation         0.000473
Name: proportion, dtype: float64

--- Distribución de deposit_type ---
Dataset completo:
deposit_type
No Deposit    0.876464
Non Refund    0.122179
Refundable    0.001357
Name: proportion, dtype: float64

Duplicados (keep=False):
deposit_type
No Deposit    0.641678
Non Refund    0.356430
Refundable    0.001892
Name: proportion, dtype: float64

--- Tasa de cancelación (is_canceled) ---
Dataset completo : 0.3704
Duplicados (keep=False): 0.5839


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

a) Las filas duplicadas corresponden casi con certeza a errores de registro (como cargas repetidas de archivos o reenvíos de formularios en los sistemas hoteleros) y no a reservas distintas. Dada la coincidencia exacta en las 32 columnas (incluyendo fechas, tarifas y estatus) en más de una cuarta parte de los datos, la probabilidad de que sean clientes distintos con itinerarios perfectamente idénticos es estadísticamente nula.

b) El método duplicated() trata todos los valores NaN (presentes en agent y company) como iguales entre sí. En este contexto, esto es correcto, ya que dos reservas directas sin agencia ni compañía intermediaria que coincidan en el resto de los campos deben considerarse duplicadas de manera legítima.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [8]:
# Escribe tu código aquí

# Creamos hoteles_limpio eliminando duplicados exactos, conservando la primera y renumerando índice
hoteles_limpio = hoteles.drop_duplicates(keep='first', ignore_index=True)

print(f"Filas antes     : {len(hoteles)}")
print(f"Filas después   : {len(hoteles_limpio)}")
print(f"Filas eliminadas: {len(hoteles) - len(hoteles_limpio)}")

Filas antes     : 119390
Filas después   : 87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [9]:
# Escribe tu código aquí

def calcular_metricas(df):
    # Calcula reservas, tasa de cancelación y adr medio agrupado por hotel
    return df.groupby('hotel').agg(
        reservas=('is_canceled', 'size'),
        tasa_cancelacion=('is_canceled', 'mean'),
        adr_medio=('adr', 'mean')
    ).reset_index()

resumen_original = calcular_metricas(hoteles)
resumen_limpio = calcular_metricas(hoteles_limpio)

print("Resumen Original:")
print(resumen_original)
print("\nResumen Limpio:")
print(resumen_limpio)

Resumen Original:
          hotel  reservas  tasa_cancelacion   adr_medio
0    City Hotel     79330          0.417270  105.304465
1  Resort Hotel     40060          0.277634   94.952930

Resumen Limpio:
          hotel  reservas  tasa_cancelacion   adr_medio
0    City Hotel     53428          0.300386  110.985944
1  Resort Hotel     33968          0.234809   99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [10]:
# Escribe tu código aquí

# Agregamos la columna 'version' a cada resumen
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

# Unimos ambos DataFrames con pd.concat y ordenamos por hotel
comparacion = pd.concat([resumen_original, resumen_limpio], ignore_index=True).sort_values(by='hotel').reset_index(drop=True)
print(comparacion)

          hotel  reservas  tasa_cancelacion   adr_medio         version
0    City Hotel     79330          0.417270  105.304465        original
1    City Hotel     53428          0.300386  110.985944  sin duplicados
2  Resort Hotel     40060          0.277634   94.952930        original
3  Resort Hotel     33968          0.234809   99.025346  sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

a) La métrica que cambia de forma drástica es el número absoluto de reservas, con una reducción importante en el City Hotel. En contraste, las tasas relativas (como la tasa de cancelación y el adr medio) varían de manera mínima (menos del 1-2%), lo que indica que los duplicados estaban distribuidos de forma homogénea en los valores numéricos y categóricos, aunque distorsionaban por completo los volúmenes totales.

b) Entrenar un modelo predictivo para is_canceled con los datos originales provocaría sobreajuste (overfitting), ya que el algoritmo aprendería patrones repetidos dándoles un peso artificialmente elevado. Asimismo, si un registro duplicado quedara en el conjunto de entrenamiento y su copia exacta en el conjunto de prueba, se generaría fuga de datos (data leakage), inflando artificialmente las métricas de evaluación del modelo.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [11]:
# Escribe tu código aquí

print("Duplicados exactos restantes:", hoteles_limpio.duplicated().sum())
print("¿Indices correctos de 0 a len-1?", (hoteles_limpio.index == range(len(hoteles_limpio))).all())
print("Total de filas en limpio :", len(hoteles_limpio))

Duplicados exactos restantes: 0
¿Indices correctos de 0 a len-1? True
Total de filas en limpio : 87396


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [12]:
# Escribe tu código aquí

# Columnas a comparar (todas excepto reservation_status_date)
cols_parcial = [c for c in hoteles_limpio.columns if c != 'reservation_status_date']

# 1. Contar filas sobrantes con este criterio parcial
dups_parciales_sobrantes = hoteles_limpio.duplicated(subset=cols_parcial, keep='first').sum()
print(f"Filas sobrantes con duplicidad parcial (excepto fecha de estado): {dups_parciales_sobrantes}")

# 2. Mostrar una muestra ordenada de estas filas
mask_parcial = hoteles_limpio.duplicated(subset=cols_parcial, keep=False)
cols_muestra = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'adr', 'reservation_status', 'reservation_status_date']
print("\nMuestra de duplicados parciales:")
print(hoteles_limpio[mask_parcial].sort_values(by=['hotel', 'lead_time'])[cols_muestra].head(10))

# 3. Distribución de reservation_status en estas filas
print("\nDistribución de reservation_status:")
print(hoteles_limpio[mask_parcial]['reservation_status'].value_counts())

Filas sobrantes con duplicidad parcial (excepto fecha de estado): 245

Muestra de duplicados parciales:
            hotel  lead_time  arrival_date_year arrival_date_month  \
35572  City Hotel          9               2015          September   
35574  City Hotel          9               2015          September   
51267  City Hotel         12               2017               July   
51268  City Hotel         12               2017               July   
39425  City Hotel         14               2016              April   
39432  City Hotel         14               2016              April   
40080  City Hotel         14               2016                May   
40083  City Hotel         14               2016                May   
54170  City Hotel         14               2015            October   
54474  City Hotel         14               2015            October   

       arrival_date_day_of_month     adr reservation_status  \
35572                         10   95.00           Canceled   

**Respuesta 4.2:**

Estas filas representan actualizaciones administrativas o cambios de estatus en el ciclo de vida de la reserva registrados en fechas posteriores (por ejemplo, una reserva que pasó de confirmada a cancelada o de la que se emitió un nuevo estatus transaccional). No se deben eliminar si el objetivo es auditar la trazabilidad operativa de los cambios de estado, pero pueden excluirse si el enfoque analítico busca únicamente el estado final a fecha de llegada.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

Criterio: Duplicidad exacta en las 32 columnas del dataset hoteles.

Regla de conservación (keep): keep='first', conservando la primera aparición registrada para mantener la estabilidad del índice original y el orden temporal de carga.

Resultado: Se redujo el dataset de 119,390 a 87,396 filas, eliminando 31,994 registros duplicados exactos (26.80%).

Supuestos y limitaciones: Se asume que la coincidencia exacta en todos los campos representa redundancias de sistema. Se omitieron criterios parciales para no perder registros de modificaciones operativas válidas.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

Situación donde no convendría eliminar duplicados exactos: En bases de datos transaccionales de cajeros automáticos (ATM) o puntos de venta (POS) donde dos clientes distintos realizan exactamente la misma compra (mismo artículo, mismo monto exacto) en momentos distintos, eliminar duplicados borraría transacciones comerciales reales.

Información adicional: Un identificador único de reserva (reservation_id) y marcas de tiempo (timestamps) precisas de la creación del registro en el servidor, lo que permitiría diferenciar con certeza los reenvíos de formularios de las reservas idénticas legítimas.

Por qué duplicated().sum() == 0 no garantiza limpieza: Porque pueden existir duplicados ocultos por diferencias de formato tipográfico (espacios, mayúsculas), variaciones en tipos de datos (números almacenados como texto), o duplicados parciales donde el identificador cambia pero todos los atributos del cliente coinciden.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.